jobs/revoke_grants_by_user_criteria.go

```diff
@@ -0,0 +1,126 @@
+	UserCriteria        evaluator.Expression `mapstructure:"user_criteria"`
+	ReassignOwnershipTo evaluator.Expression `mapstructure:"reassign_ownership_to"`
+}
+
+func (h *handler) RevokeGrantsByUserCriteria(ctx context.Context, c Config) error {
+	h.logger.Info(fmt.Sprintf("starting %q job", RevokeGrantsByUserCriteria))
+
+	var cfg RevokeGrantsByUserCriteriaConfig
+	if err := c.Decode(&cfg); err != nil {
+		return fmt.Errorf("invalid config for %s job: %w", RevokeGrantsByUserCriteria, err)
+	}
+
+	iamManager := identities.NewManager(h.crypto, h.validator)
+	iamConfig, err := iamManager.ParseConfig(&cfg.IAM)
+	if err != nil {
+		return fmt.Errorf("parsing IAM config: %w", err)
+	}
+	iamClient, err := iamManager.GetClient(iamConfig)
+	if err != nil {
+		return fmt.Errorf("initializing IAM client: %w", err)
+	}
+
+	activeGrants, err := h.grantService.List(ctx, domain.ListGrantsFilter{
+		Statuses: []string{string(domain.GrantStatusActive)},
+	})
+	if err != nil {
+		return fmt.Errorf("listing active grants: %w", err)
+	}
+
+	grantsForUser := map[string][]*domain.Grant{}     // map[account_id][]grant
+	grantsOwnedByUser := map[string][]*domain.Grant{} // map[owner][]grant
+	uniqueUserEmails := map[string]bool{}             // map[account_id]bool
+	for _, g := range activeGrants {
+		if g.AccountType == domain.DefaultAppealAccountType {
+			// collecting grants for individual users
+			grantsForUser[g.AccountID] = append(grantsForUser[g.AccountID], &g)
+			uniqueUserEmails[g.AccountID] = true
+		} else if g.Owner != domain.SystemActorName {
+			// collecting other grants owned by the user
+			grantsOwnedByUser[g.Owner] = append(grantsOwnedByUser[g.AccountID], &g)
+		}
+	}
+
+	for email := range uniqueUserEmails {
+		user, err := iamClient.GetUser(email)
+		if err != nil {
+			h.logger.Error("getting user details from identity manager", "email", email, "error", err)
+			continue
+		}
+		userDetails, ok := user.(map[string]interface{})
+		if !ok {
+			return fmt.Errorf("parsing user details: expected a map[string]interface{}, got %T instead; value is %q", user, user)
+		}
+		evaluationParams := map[string]interface{}{
+			"user": userDetails,
+		}
+
+		// evaluating user criteria
+		criteriaEvaluation, err := cfg.UserCriteria.EvaluateWithVars(evaluationParams)
+		if err != nil {
+			return fmt.Errorf("evaluating user_criteria: %w", err)
+		}
+		criteriaSatisfied, ok := criteriaEvaluation.(bool)
+		if !ok {
+			return fmt.Errorf("invalid type for user_criteria evaluation result: expected boolean, got %T; value is %q", criteriaEvaluation, criteriaEvaluation)
+		} else if !criteriaSatisfied {
+			continue
+		}
+
+		// revoking grants with account_id == email
+		revokeGrantsFilter := domain.RevokeGrantsFilter{
+			AccountIDs: []string{email},
+		}
+		h.logger.Info("revoking grants", "account_id", email)
+		// revokedGrants, err := h.grantService.BulkRevoke(ctx, revokeGrantsFilter, domain.SystemActorName, "Revoked due to user deactivated")
```

Comment:
maybe forgot to remove this?

Reply:
@mabdh it has been updated, please check the latest changes